backends/xsmm/ceed-xsmm-tensor.c

```diff
@@ -73,27 +66,112 @@ static int CeedTensorContractApply_Xsmm(CeedTensorContract contract, CeedInt A,
                                         CeedInt B, CeedInt C, CeedInt J,
                                         const CeedScalar *restrict t,
                                         CeedTransposeMode tmode,
-                                        const CeedInt Add,
+                                        const CeedInt add,
                                         const CeedScalar *restrict u,
                                         CeedScalar *restrict v) {
-  CeedInt blksize = 8;
+  int ierr;
+  CeedInt blksize = 8, ind, nelem;
+  CeedTensorContract_Xsmm *impl;
+  ierr = CeedTensorContractGetData(contract, (void *)&impl); CeedChk(ierr);
 
   if (C % blksize)
-    CeedTensorContract_Xsmm_Serial(contract, A, B, C, J, t, tmode, Add, u, v);
+    nelem = 1;
+  else
+    nelem = blksize;
+
+  // Get kernel index
+  if (impl->tensorbasis)
+    ind = CeedGetXsmmInd_Tensor(nelem, add?1:0, tmode==CEED_TRANSPOSE?1:0, B, C,
+                                J, impl->dim);
+  else
+    ind = CeedGetXsmmInd_NonTensor(nelem, add?1:0, impl->P, impl->Q, B, J);
+
+  // Run kernel or fallback to default implementation
+  if (C != 1 && ind < impl->numkernels && ind >= 0 && impl->kernels[ind])
+    for (CeedInt a=0; a<A; a++)
+      impl->kernels[ind](&u[a*B*C], &t[0], &v[a*J*C], NULL, NULL, NULL);
   else
-    CeedTensorContract_Xsmm_Blocked(contract, A, B, C, J, t, tmode, Add, u, v);
+    CeedTensorContract_Xsmm_Default(contract, A, B, C, J, t, tmode, add, u, v);
 
   return 0;
 }
 
 static int CeedTensorContractDestroy_Xsmm(CeedTensorContract contract) {
+  int ierr;
+  CeedTensorContract_Xsmm *impl;
+  ierr = CeedTensorContractGetData(contract, (void *)&impl); CeedChk(ierr);
+  ierr = CeedFree(&impl->kernels); CeedChk(ierr);
+  ierr = CeedFree(&impl); CeedChk(ierr);
+
   return 0;
 }
 
-int CeedTensorContractCreate_Xsmm(CeedTensorContract contract) {
+int CeedTensorContractCreate_Xsmm(CeedBasis basis,
+                                  CeedTensorContract contract) {
   int ierr;
   Ceed ceed;
   ierr = CeedTensorContractGetCeed(contract, &ceed); CeedChk(ierr);
+  CeedTensorContract_Xsmm *impl;
+  ierr = CeedCalloc(1, &impl); CeedChk(ierr);
+
+  // Set up pointers to kernels
+  ierr = CeedBasisGetTensorStatus(basis, &impl->tensorbasis); CeedChk(ierr);
+  if (impl->tensorbasis) {
+    ierr = CeedBasisGetNumNodes1D(basis, &impl->P); CeedChk(ierr);
+    ierr = CeedBasisGetNumQuadraturePoints1D(basis, &impl->Q); CeedChk(ierr);
+    ierr = CeedBasisGetDimension(basis, &impl->dim); CeedChk(ierr);
+    // Set up kernel pointer array
```

Comment:
The kernel set up feels a bit hard to read to me? Is it at all understandable to someone who didn't write it?